Eclio lgr

tskille opened the pull request:
The following PR includes updates to the EclIO classes **EGrid** and **ERst**.

The main purpose of these updates are to enable parsing and extracting data from Eclipse files from models using LGRs.

A new class **EInit** has been made for parsing INIT files. Init files without LGRs can easily be parsed with the base EclIO class **EclFile**. However, Init files from models using LGRs will have multiple instances of the same keywords and this new class will make it much easier to extract data from the init file.

akva2 commented on opm/io/eclipse/EclFile.hpp:
```diff
@@ -38,12 +38,12 @@ class EclFile
-    void loadData(const std::string& arrName);         // load all arrays with array name equal to arrName
-    void loadData(int arrIndex);                // load data based on array indices in vector arrIndex
-    void loadData(const std::vector<int>& arrIndex);   // load data based on array indices in vector arrIndex
+    void loadData() const;                                  // load all data
```
this makes little sense. this is a loader, why do you want to make them const? seems just confusing and misleading.

akva2 commented on opm/io/eclipse/EclFile.hpp:
```diff
@@ -38,12 +38,12 @@ class EclFile
+    void loadData(const std::vector<int>& arrIndex) const;  // load data based on array indices in vector arrIndex
 
-    void clearData()
+    void clearData() const
```
likewise. this clears data, it implies writes and should not be const

akva2 commented on opm/io/eclipse/EclFile.hpp:
```diff
@@ -110,10 +110,10 @@ class EclFile
 
-    void loadBinaryArray(std::fstream& fileH, std::size_t arrIndex);
-    void loadFormattedArray(const std::string& fileStr, std::size_t arrIndex, int64_t fromPos);
+    void loadBinaryArray(std::fstream& fileH, std::size_t arrIndex) const;
```
this writes to class members. constness is not warranted, and very confusing.

akva2 commented on src/opm/io/eclipse/EInit.cpp:
```diff
@@ -0,0 +1,161 @@
+
+EInit::EInit(const std::string &filename) : EclFile(filename)
+{
+    std::string lgrname = "";
```
strings are blank on default initialization.

tskille replied:
Thanks was not sure about this. Changed to default initialization.

akva2 commented on src/opm/io/eclipse/EInit.cpp:
```diff
@@ -0,0 +1,161 @@
+{
+    if (grid_name == "global"){
+        if (global_array_index.count(name) == 0)
+            OPM_THROW(std::runtime_error, "Map key not found in global_array_index");
```
which key should be in error message.

tskille replied:
agree, throw message are improved.

akva2 commented on src/opm/io/eclipse/EInit.cpp:
```diff
@@ -0,0 +1,161 @@
+        int lgr_index = get_lgr_index(grid_name);
+
+        if (lgr_array_index[lgr_index].count(name) == 0)
+            OPM_THROW(std::invalid_argument, "Map key not found in lgr_array_index");
```
same. include key name.

tskille replied:
throw message are improved.

akva2 commented on src/opm/io/eclipse/EInit.cpp:
```diff
@@ -0,0 +1,161 @@
+
+int EInit::get_lgr_index(const std::string& grid_name) const
+{
+    if (std::find(lgr_names.begin(), lgr_names.end(), grid_name) == lgr_names.end()) {
```
store the iterator, to avoid double search.

tskille replied:
Done

akva2 commented on opm/io/eclipse/EGrid.hpp:
```diff
@@ -50,14 +51,48 @@ class EGrid : public EclFile
     int activeCells() const { return nactive; }
     int totalNumberOfCells() const { return nijk[0] * nijk[1] * nijk[2]; }
 
+    void load_grid_data() const;
```
why const?

akva2 commented on opm/io/eclipse/EGrid.hpp:
```diff
@@ -50,14 +51,48 @@ class EGrid : public EclFile
     int totalNumberOfCells() const { return nijk[0] * nijk[1] * nijk[2]; }
 
+    void load_grid_data() const;
+    void load_nnc_data() const;
```
why const?

akva2 commented on src/opm/io/eclipse/EGrid.cpp:
```diff
@@ -28,43 +30,188 @@
+
+            if (array_name[n] == "COORD")
+                coord_array_index= n;
+            if (array_name[n] == "ZCORN")
```
it cannot be all of these, use else to avoid unnecessary string comparisons.

tskille replied:
you are correct. using else if.

akva2 commented on src/opm/io/eclipse/EGrid.cpp:
```diff
@@ -28,43 +30,188 @@
+            int init_nactive = init.activeCells(m_grid_name);
+
+            if ((init_dims != nijk) or (init_nactive != nactive)){
+                OPM_THROW(std::invalid_argument, "inconsistent grid dimension in init file");
```
found blah, expected blah. also 'or' is for wusses ;p

tskille replied:
Throw message improved. Also split original if statement (including or) into two different if statements. 

NNC definition and NNC transmissibilities store in different files (egrid and init file). I would really like to be sure that these are compatible.

akva2 commented on src/opm/io/eclipse/EGrid.cpp:
```diff
@@ -28,43 +30,188 @@
+            auto trans_data = init.getInitData<float>("TRANNNC", m_grid_name);
+
+            if (trans_data.size() != nnc1_array.size())
+                OPM_THROW(std::invalid_argument, "inconsistent size of array TRANNNC in init file");
```
values blah and blah.

tskille replied:
throw message improved

akva2 commented on src/opm/io/eclipse/EGrid.cpp:
```diff
@@ -131,48 +278,61 @@ void EGrid::getCellCorners(const std::array<int, 3>& ijk,
                            std::array<double,8>& Y,
                            std::array<double,8>& Z) const
 {
+    if (coord_array.size() == 0)
```
.empty()

tskille replied:
thanks, changed

akva2 commented on opm/io/eclipse/ERst.hpp:
```diff
@@ -37,43 +38,73 @@ class ERst : public EclFile
     bool hasReportStepNumber(int number) const;
+    bool hasLGR(const std::string& gridname, int reportStepNumber) const;
+
+    void loadReportStepNumber(int number) const;
```
why const?

akva2 commented on src/opm/io/eclipse/ERst.cpp:
```diff
@@ -101,36 +105,66 @@ std::vector<EclFile::EclEntry> ERst::listOfRstArrays(int reportStepNumber)
+        if ((array_name[n] != "SEQNUM") && (array_name[n] != "LGR"))
+            list.emplace_back(array_name[n], array_type[n], array_size[n]);
+
+    } while ((array_name[n] != "SEQNUM") && (array_name[n] != last_array_name) && (n < static_cast<int>((array_name.size() -1 ))));
```
subtraction should be done after cast.

tskille replied:
Done

akva2 reviewed (commented):
Mostly small stuff, but letting the hack of loading in getters bleed into public interface is a no go. If you have to do these nasty things, you should hide it in your implementation and use const_cast at call sites.

tskille commented:
Thanks @akva2 for your comments. Making all of these member functions in EclFile const seemed as a good idea, but I realize that it wasn't. I have reverted the entire commit "EclFile updates". I also removed all mutable definitions and remove all const needed for compiling of ERst, EInit and EGrid. 

I have also re-based this with lates master. Hopefully this should be ready now.

bska commented:
jenkins build this please

tskille commented:
Sorry, forgot to update python bindings. Now fixed, hopefully green on next build.

bska commented:
jenkins build this please

akva2 commented:
thanks for the updates. i'm merging.

bska commented on src/opm/io/eclipse/EGrid.cpp:
```diff
@@ -28,44 +30,203 @@
+                nijk[1] = gridhead[2];
+                nijk[2] = gridhead[3];
+
+                m_radial = gridhead[26] > 0 ? true: false;
```
Please note that this (accessing `gridhead[26]`)  combined with the unit test's [definition](https://github.com/OPM/opm-common/blob/4c48f9bcf91f91974c6cf505b5f4129814d2e589/tests/test_EclRegressionTest.cpp#L315)
```C++
std::vector<int> gridhead = {1,2,3,2,0,0};
```
(`gridhead.size() == 6`) generates a guaranteed out-of-bounds indexing error.

I tried to fix this locally in the obvious way&mdash;extending the size of `gridhead` in `test_EclRegressionTest.cpp:gridCompare`&mdash;but that only led to more unit test failures.

What is the correct way to fix the out-of-bounds error here?

tskille replied:
Thanks @bska  for spotting this error/bug. I think that the best way to fix this is to initialize the variable m_radial to false in addition to adding a check for size of gridhead

```c++
if (gridhead.size() > 26)
    m_radial = gridhead[26] > 0 ? true: false;
```

I can add a PR to fix this if you agree with this solution
